chore: Updating config readme (#288)
skip-mev/slinky: oracle/config/README.md

# Configurations

> > This readme overviews how to configure the oracle side-car as well as how to hook it up to your application. To see an example of a properly configured oracle-side car, please visit the [local config](./../../config/local) files - `oracle.json` and `market.json`. To see an example of a properly configured application, please visit the test application's [app.toml](./../../tests/simapp/slinkyd/testappd/root.go) generation code. Otherwise, please read on to learn how to configure the oracle side-car and application.

Reviewer: ```suggestion
> This readme overviews how to configure the oracle side-car as well as how to hook it up to your application. To see an example of a properly configured oracle-side car, visit the [local config](./../../config/local) files - `oracle.json` and `market.json`. To see an example of a properly configured application, please visit the test application's [app.toml](./../../tests/simapp/slinkyd/testappd/root.go) generation code. Otherwise, read on to learn how to configure the oracle side-car and application.
```

>

> > To generate custom side-car configs, please use the `slinky-config` binary after running `make build`. Run `./build/slinky-config --help` to determine the relevant fields that can be set.

Reviewer: ```suggestion
> To generate custom side-car configs, use the `slinky-config` binary after running `make build`. Run `./build/slinky-config --help` to determine the relevant fields that can be set.
```

> 

> > Validator's running on a network that support's Slinky **must** run the oracle side-car and configure it into their application. Non-validator's can configure their oracle config's to be disabled, and the oracle side-car will not be run.

Reviewer: ```suggestion
> Validators running on a network that support's Slinky **must** run the oracle side-car and configure it into their application. Non-validators can configure their oracle config's to be disabled, and the oracle side-car will not be run.
```

>
> <div align="center">
> 
> | Type | Oracle/Market Config | Oracle Metrics | App Metrics |
> |----------|:--------:|---------:|--------:|
> | Validator     | **Required**     |    **Recommended**   | **Recommended** |
> | Non-Validator | **Optional**     |    **Optional**   | **Optional** |
> </div>

All oracle configurations are broken down into three files:

1. **Oracle side-car configuration (`oracle.json`):** This contains the data provider's that are utilized, how often they should be polled, and a variety of other configurations for API and web socket providers.

> 2. **Market side-car configuration (`market.json`):** This contains the desired markets that the side-car will fetch prices for. NOTE: It is recommended that this file is **NOT** modified nor created by validators. This file is typically provided by the chain that the oracle supports.

Reviewer: ```suggestion
2. **Market side-car configuration (`market.json`):** This contains the desired markets that the side-car will fetch prices for. NOTE: It is recommended that this file is **NOT** modified nor created by validators. This file is typically provided by the chain that the side-car supports.
```

# Oracle Side-Car Configuration

The `oracle.json` file is the configuration file that is consumed by the oracle side-car. Note that in most cases, this should **NOT** be custom made by validators - unless specified otherwise. A predefined oracle side car configuration should be provided by the chain that the oracle supports. This file contains:

* The desired data providers to be utilized i.e. Coinbase, Binance, etc.
* Metrics instrumentation.
* API & WebSocket configurations for each provider.

> In some cases, validators must configure the market map provider into their `oracle.json`. The market map provider is a special provider that provides the desired markets that the oracle should fetch prices for. This is particularly useful for chains that have a large number of markets that are constantly changing. The market map provider allows the oracle to be updated with new markets without needing to restart the side-car. **Please check the relevant chain's documentation & channels to determine if you need to configure the market map provider.**

Reviewer: ```suggestion
In some cases, validators must configure the market map provider into their `oracle.json`. The market map provider is a special provider that provides the desired markets that the oracle should fetch prices for. This is particularly useful for chains that have a large number of markets that are constantly changing. The market map provider allows the side-car to be updated with new markets without needing to restart the side-car. **Please check the relevant chain's documentation & channels to determine if you need to configure the market map provider.**
```

```go
type OracleConfig struct {
	UpdateInterval time.Duration    `json:"updateInterval"`
	MaxPriceAge    time.Duration    `json:"maxPriceAge"`
	Providers      []ProviderConfig `json:"providers"`
	Production     bool             `json:"production"`
	Metrics        MetricsConfig    `json:"metrics"`
	Host           string           `json:"host"`
	Port           string           `json:"port"`
}
```

## UpdateInterval

> This field is utilized to set the interval at which the oracle will aggregate price feeds from price providers.

Reviewer: ```suggestion
This field is utilized to set the interval at which the side-car will aggregate price feeds from price providers.
```

## MaxPriceAge

> This field is utilized to set the maximum age of a price that the oracle will consider when aggregating prices. If a price is older than this value, the oracle will not consider it when aggregating prices.

Reviewer: ```suggestion
This field is utilized to set the maximum age of a price that the oracle will consider when aggregating prices. If a price is older than this value, the side-car will not consider it when aggregating prices.
```

## Providers

This field is utilized to set the list of providers that the oracle will fetch prices from. A given provider's configuration is composed of:

* An API configuration that defines the various API configurations that the oracle will use to fetch prices from the provider.
* A WebSocket configuration that defines the various WebSocket configurations that the oracle will use to fetch prices from the provider.
* A type that defines the type of provider - specifically this is currently either a price or market map provider. Price providers supply price feeds for a given set of markets, while a market map provider supplies the desired markets that need to be fetched. Market map providers allow the side-car to be updated with new markets without needing to restart the side-car.

> > Note: Typically only one of either the API or websocket config is required. However, some providers may require both. Please read the provider's documentation to learn more about how to configure the provider. Each provider provides sensible defaults for the API and WebSocket configurations that should be used for most cases. This should be modified with caution.

Reviewer: ```suggestion
> Note: Typically only one of either the API or WebSocket config is required. However, some providers may require both. Please read the provider's documentation to learn more about how to configure the provider. Each provider provides sensible defaults for the API and WebSocket configurations that should be used for most cases. This should be modified with caution.
```

> > Note: Typically only one of either the API or WebSocket config is required. However, some providers may require both. Please read the provider's documentation to learn more about how to configure the provider. Each provider provides sensible defaults for the API and WebSocket configurations that should be used for most cases. This should be modified with caution.

Reviewer: ```suggestion
> Note: Typically only one of either the API or WebSocket config is required. However, some providers may require both. Read the provider's documentation to learn more about how to configure the provider. Each provider provides sensible defaults for the API and WebSocket configurations that should be used for most cases. This should be modified with caution.
```
